databuilder/contracts/contracts.py

```diff
@@ -1,33 +1,27 @@
-from . import types
-from .base import Column, TableContract
-from .constraints import FirstOfMonthConstraint, NotNullConstraint, UniqueConstraint
+import datetime
 
+from databuilder.query_language import EventFrame, PatientFrame, Series, construct
 
-class PatientDemographics(TableContract):
-    """Provides demographic information about patients."""
+from .constraints import FirstOfMonthConstraint, NotNullConstraint
 
-    _name = "patient_demographics"
 
-    patient_id = Column(
-        type=types.PseudoPatientId(),
-        description=(
-            "Patient's pseudonymous identifier, for linkage. You should not normally "
-            "output or operate on this column"
-        ),
-        constraints=[NotNullConstraint(), UniqueConstraint()],
-    )
-    date_of_birth = Column(
-        type=types.Date(),
+@construct
+class patient_demographics(PatientFrame):
+    """Provides demographic information about patients."""
+
+    date_of_birth = Series(
+        datetime.date,
         description="Patient's year and month of birth, provided in format YYYY-MM-01.",
         constraints=[NotNullConstraint(), FirstOfMonthConstraint()],
     )
-    sex = Column(
-        type=types.Choice("female", "male", "intersex", "unknown"),
+    sex = Series(
+        str,
+        choices=["female", "male", "intersex", "unknown"],
```

Comment:
Yes, this is a good point. I was thinking that if we went down this road we'd need distinct types for each set of categories, and that gets complicated fast; but you're quite right that there could be value in having a `Categorical` type distinct from `str` even if that doesn't itself track the categories (that _is_ your suggestion, right?).

I don't think we can do this immediately though, because it would need changes in ehrQL along the lines of the implicit string-to-date code in order to avoid users having to write `patients.sex == Categorical("F")`. So I think in the short term leaving it as `str` but rendering it into the docs as `Categorical` is probably the least worst option.

Reply:
> if we went down this road we'd need distinct types for each set of categories

That actually hadn't occurred to me, but under exactly the current scheme you're quite right, of course.

We're using `type` for a bunch of different things here:

* constrain the types of objects that can be use in ehrQL
* constrain the values that can be used in ehrQL (for example for dates-as-strings)
* specify conversions in translating to SQL
* (presumably, later) validate types actually found in the database

And of course there's a lot of overlap with `constraints`.

The more I think about it the more tricky this seems to be. I'm pretty much convinced that the current model is not correct, but it needs more thinking to suggest a better alternative.

I'm inclined to suggest that there should be a single object which describes all constraints (types and values) on a given column -- or maybe that it should be split into two, one for ehrQL and one for the database.

> I think in the short term leaving it as str but rendering it into the docs as Categorical is probably the least worst option.

I agree, and I think this would be good enough for now with the addition (I think) that the `choices` argument should be folded into `constraints`.